### ISSUE TITLE
Refactor FileMonitor because of inconsistent behavior when heavy log activity

### ISSUE THREAD

**Issue by mdengra** [wazuh/wazuh-qa#1602]
## Description
Some tests fail randomly because a certain message that the test expected to be written was not detected in the log file. 

After multiple tests, it appears that these errors originate in the `monitoring.py` module of the QA framework, specifically in the `FileMonitor` class, which is the one that finally checks if the related logs have been written. 

These logs are written to the log file correctly, but the `FileMonitor` class is not able to detect them when a large number of messages are written to the log file (**log flooding**). To verify this I have disabled the `sca` module, which generates a large number of logs and the tests involved have been running correctly.

### Package Details
 
|  Jenkins branch  | QA branch   |  Version 	|   Revision | 
|- |- |- |- |
|  4.2 |   1516-4.2.0-full-green | 4.2  |  1.1493 | 

Type | Format | Architecture | Version | Revision | Tag | File name
-- | -- | -- | -- | -- | -- | --
Server | rpm | x86_64 | 4.2.0 | 40207 | v4.2.0-rc8 | wazuh-manager-4.2.0-1.1493.x86_64.rpm

### Environment 

Provider| Box| OS| CPU| Memory | 
--|--|--|--|--|
Vagrant | centos/8 | CentOS 8 | 2 | 1024 |

### Setup local_internal_options.conf
```
wazuh_modules.debug=2
monitord.rotate_log=0
```

### Case:

|  Type | Description  |
|--|--|
|Test Path | test_vulnerability_detector/test_feeds/msu/test_missing_fields_msu_feed.py
| Error message |  "Could not find the message: 'Microsoft Security Update' feed finished successfully"

## Actual Behavior
Here you can see the instant when the test fails (test run using pytest **-sv** options):
```
test_vulnerability_detector/test_feeds/msu/test_missing_fields_msu_feed.py::test_invalid_msu_feed[MSU_configuration-missing: restart_required] 2021/07/14 08:22:34 wazuh-modulesd[60743] debug_op.c:70 at _log(): DEBUG: Logging module auto-initialized
2021/07/14 08:22:34 wazuh-modulesd[60743] main.c:76 at main(): DEBUG: Wazuh home directory: /var/ossec
2021/07/14 08:22:34 wazuh-modulesd[60743] wmodules-osquery-monitor.c:78 at wm_osquery_monitor_read(): DEBUG: Logpath read: /var/log/osquery/osqueryd.results.log
2021/07/14 08:22:34 wazuh-modulesd[60743] wmodules-osquery-monitor.c:84 at wm_osquery_monitor_read(): DEBUG: configPath read: /etc/osquery/osquery.conf
2021/07/14 08:22:34 wazuh-modulesd[60743] wmodules-vuln-detector.c:609 at wm_vuldet_read_provider(): DEBUG: Added msu feed. Interval: 3600s | Multi path: '/home/vagrant/wazuh-qa/tests/integration/test_vulnerability_detector/test_feeds/msu/../../data/feeds/custom_msu.json$' | Multi url: 'none' | Update since: 0 | Timeout: 300s
2021-07-14 08:22:54,456 - wazuh_testing - ERROR - Could not find the message: 'Microsoft Security Update' feed finished successfully
2021-07-14 08:22:54,456 - wazuh_testing - ERROR - Results accumulated: 0
2021-07-14 08:22:54,456 - wazuh_testing - ERROR - Results expected: 1
FAILED2021-07-14T08:22:56.1626250976
```

The use case starts running at `08:22:34`:
`2021/07/14 08:22:34 wazuh-modulesd[60743] main.c:76 at main(): DEBUG: Wazuh home directory: /var/ossec`

And ends at `08:22:54` due to timeout (20s):
`2021-07-14 08:22:54,456 - wazuh_testing - ERROR - Could not find the message: 'Microsoft Security Update' feed finished successfully`

However, by analyzing the `ossec.log` file, you can see that the message mentioned by the error has been written:
`2021/07/14 08:22:34 wazuh-modulesd:vulnerability-detector[60743] wm_vuln_detector.c:4077 at wm_vuldet_check_feed(): INFO: (5430): The update of the 'Microsoft Security Update' feed finished successfully.`

As mentioned above, this behavior does not occur if modules that write a large number of messages to the log file, such as the `sca` module, are disabled.

## Expected Behavior
The test pass because FileMonitor detects the expected message.

## Note

| Environment Setup | Test Log   | Ossec Setting |
|--|--|--|
[Steps to create the testing environment](https://github.com/wazuh/wazuh-qa/files/6816104/environment_setup.txt) | [test_extra_fields_msu_feed_local.log](https://github.com/wazuh/wazuh-qa/files/6816052/test_extra_fields_msu_feed_local.log) |[ossec.conf.txt](https://github.com/wazuh/wazuh-qa/files/6816054/ossec.conf.txt)|

### Issues related to the problem described above:
- https://github.com/wazuh/wazuh-qa/issues/1542
- https://github.com/wazuh/wazuh-qa/issues/1547
- https://github.com/wazuh/wazuh-qa/issues/1548
- https://github.com/wazuh/wazuh-qa/issues/1553

**Comment by mdengra** [comment]
Issues related to the problem described above:
  - https://github.com/wazuh/wazuh-qa/issues/1547
  - https://github.com/wazuh/wazuh-qa/issues/1548

**Comment by MizugorouZ** [comment]
I have created an script with the help of @Rebits ir order to find the origin of the problem. The script itself has 2 parts: the first one is create a file of 1.000.000 lines, in which each line is "Line {number}". The second part of the script uses the FileMonitor class, to find a random line from the file with 1.000.000. The result is that FileMonitor cannot find the line due to timeout error. After changing several values I got the following results:
- Wtih 1.000.000 lines and 10 seconds of timeout, the last line to read was in the range of 2800-3200
- Wtih 1.000.000 lines and 60 seconds of timeout, the last line to read was about ~25000
- With 100.000 lines and 60 seconds of timeout, the last line to read was about ~33000

Lines are generated well, but FileMonitor cannot reach them due to timeouts and large amount of lines. Next thing I will do is execute the tests with the settings above.

**Comment by Rebits** [comment]
After some research, we have concluded the following facts about `FileMonitor`:

 - `FileMonitor` does detect every message generated in the file. 
 - Truncate file does not affect to FileMonitor log detection.
 - `FileMonitor` QueueMonitor attribute fails to monitor in time every line stored but the `FileTailer` in time.

`FileMonitor` has a `FileTailer` attribute that is initialised at object creation. This object will store every log message in a stack. This tailing is fast enough to monitor wazuh-qa testing logs. However, the `FileMonitor` start method allows the `update_position` attribute. This option is used in order to do not pop FileTailer stored messages. This is implemented by the peek method as follows:
https://github.com/wazuh/wazuh-qa/blob/ef1723d37835705eeceeb1a7f2abc0c1ef4ceec1/deps/wazuh_testing/wazuh_testing/tools/monitoring.py#L486
```
       aux_queue = queue.Queue()
        aux_queue.queue = copy(self.queue)
        for _ in range(position):
            aux_queue.get(*args, **kwargs)
        return aux_queue.get(*args, **kwargs)
```

If this is method is used as the `FileTailer` stack is increasing, more time is required to obtain specified elements. This is incredibly costly when all modules are activated and finally causes false-positive cases.

Regarding `test_missing_fields_msu_feed.py` tests specifically:

- It should be desirable to refactor fixtures in order to make them as modular as possible. The fixture `remove_field_feed` should only include the logic related to the fixture name, so, restart the `modulesd` daemon and truncate the file should be moved to another fixture.
- We have noticed that using this approach make the test works without a problem even using update_position to false and with all modules activated. A further investigation is necessary to determine the causes of this issue.
